### ISSUE TITLE
EXPLAIN SYNTAX + UNION DISTINCT in subquery crash

### ISSUE THREAD

**Issue by ucasFL** [ClickHouse/ClickHouse#20876]
How to reproduce:

```sql
EXPLAIN SYNTAX
SELECT x
FROM 
(
    SELECT 1 AS x
    UNION DISTINCT
    SELECT 1
)

Query id: e777bc02-5d1b-4d49-ba89-4ebe31172387


Exception on client:
Code: 32. DB::Exception: Attempt to read after eof: while receiving packet from localhost:9000

```

**Comment by ucasFL** [comment]
Hi, @KochetovNicolai, I'm trying to fix this, but I'm not very familiar with the explain query implementation......

**Comment by filimonov** [comment]
Related to https://github.com/ClickHouse/ClickHouse/issues/21170

**Comment by ucasFL** [comment]
> Related to #21170

Both of them will be fixed in #21246.
